### DISCUSSION TITLE
Ability to delete problematic OCA Files

### DISCUSSION THREAD

**Discussion by carlyh-micb** [the-human-colossus-foundation/oca-spec#29] [General]
Paul and I have been discussing a use case that might be problematic and we wanted to ask what the solution could be.

As OCA Repo exists right now, we could have malicious (or clueless) users make mistakes in schema creation. Within the text fields of meta - name and description, or within an attribute name, or information or label overlays, it is possible for someone to create problematic text. For example, in an information overlay I could write "Carly's SIN number is xxxxx". Mistake or not, this information would be available permanently within the repository as it exists right now.
What methods would we have to address removal or hiding of this problematic information? Is there a way to create support for a governance mechanisms to temporarily (or permanently) block this information?

**Comment by blelump** [comment]
As you mentioned, the right to data erasure is the governance topic. Leaving the underlying reason behind why "corrupted" information was pushed into the Repo, erasing or hiding such information is challenging. Therefore, to limit the corrupted information in the Repo, we must employ different mechanisms: reputation and the whole authenticity layer. Digitally signed information allows us to verify its origin unambiguously. We furthermore can impose (via governance) that an instance of OCA Repo accepts digitally signed information from X or Y and no one else. X or Y act as validators of the information, so what's in the Repo is digitally stamped by them and verified by a human, primarily according to law and regulations.

It is worth noting that OCA Repos can act as a reputation source, where multiple repos interact and gradually enrich the information. In other words, Repo instances may mimic the human layers, i.e., a university has departments, etc. Before information can be pushed into the university-level Repo, it must go through the department repo first. All these mechanisms limit the space for "corrupted" information.

EDIT:

Another option is introducing "human crawlers" to review the Repo contents and mark "corrupted" content according to the governance, i.e., by their digest. This approach, however, is constantly exposed to malicious users because it assumes the information filtering is a post-process rather than a pre-process, as in the above paragraphs. A malicious user can change anything to the Bundle to alter the digest and repeatedly upload Bundles with corrupted information.

**Reply by carlyh-micb** [reply]
The Oxford Dictionary definition of "ward" is the state of being in the care of a guardian. Therefore, a "wardable repository" refers to a repository entrusted to a object guardian to protect and defend against unauthorized access, modification, misuse, or safeguarding against corrupt SAIDs. The object guardian is responsible for preserving the wardable repository ensuring OCA integrity and security in the digital space. By appointing a guardian to a repository, governance administrations can establish measures to prevent unauthorized objectual actions and maintain the repository's contents' confidentiality, availability, and integrity.

We should have strong governance to avoid putting bad data into an OCA Repo in the first place. Currently there is no authentication layer implemented and this still requires considerable development and probably won't be available for a while. 
However, mistakes will be inevitable because humans do not always act perfectly. Also, something that wasn't PII or risky in the past may become problematic in the future.
In addition to a pre-processing governance, we should support post-processing governance at the same time.

We propose the concept of a wardable repository under the control of the object guardian who, according to governance rules, protects the repository contents.

One tool that we could add to a repository functionality is the ability to filter results from the repository via a blacklist of problematic SAIDs. In the beginning this can simply be a config file that can be updated manually by the owner of the repository until more complete authentication has been implemented.

Here is a use case where 'bad' is an example of a problematic entry.

ADD ATTR name // SAID 1
ADD ATTR age // SAID 2
ADD ATTR bad // SAID 3
ADD INFORMATION en ATTR name "the persons name" // SAID 4
ADD INFORMATION en ATTR bad "hahahaha" // SAID 5
ALTER ATTR bad "good" // SAID 6

FROM SAID 4
ADD META name New Schema from bad data // SAID 5a
ADD META description This is FROM'd a schema with data that will be blacklisted // SAID 6a

FROM SAID 2
ADD ATTR good // SAID 3b
ADD META name New schema from good data // SAID 4b

Now the object guardian is informed or, or finds the bad information. She adds SAID 3 (and thus all children SAIDs) to the Repo filter/blacklist.
Being in the blacklist means that the data isn't removed from the Repo. But the SAIDs don't show up in any search result and they cannot be FROM'd.
Blacklisting SAID 3 means SAID 4, 5, 6, 5a and 6a are also added to the blacklist.

SAID 3b and 4b (as well as SAID 1 and 2) are not blacklisted and continue to show in search results and can be FROM'd.

The object guardian of the repository is not going to be able to check everything that is published. Therefore we propose a feedback mechanism where problematic SAIDs can be reported for review by the object guardian.

In addition, the data catalogue idea that is currently under development can also be automatically updated and, post-review, remove catalogue entries that refer to blacklisted SAIDs.

**Reply by blelump** [reply]
To understand the issue, we have first to introduce how DAGs work. Internally, everything is stored as one tremendous DAG. For example
```
ADD ATTR name // SAID 1
ADD ATTR age // SAID 2
ADD ATTR bad // SAID 3
```
and
```
ADD ATTR name // SAID 1
ADD ATTR age // SAID 2
ADD ATTR good // SAID 3
ADD ATTR alsogood // SAID 4
```
are the same up to SAID3. Blacklisting SAID3 blacklists everything that comes after SAID3, even if it makes sense. So any ever-created capture base that contains attrs `name`, `bad`, and `age` would be automatically excluded. Any more granular level of control doesn't solve this, as blacklisting any SAID X and saying all its children are also excluded ends up with the same issue, not to mention that more granular control requires a lot more SAIDs to be blacklisted (we're deeper in the DAG and more branches can pop up).

Therefore blacklisting SAIDs and all their children is not an option.

Note we do not internally care about the instruction ordering that allows us to have the same result (SAID) for different orders of instructions, i.e., 

```
ADD ATTR name
ADD ATTR age
vs.
ADD ATTR age
ADD ATTR name
```
end up with the same SAID.

**Reply by carlyh-micb** [reply]
"Blacklisting SAID3 blacklists everything that comes after SAID3, even if it makes sense. So any ever-created capture base that contains attrs name, bad, and age would be automatically excluded."

Yes, this is the behaviour we are looking for. 

Do you have an alternative solution in mind? One that doesn't require only ensuring people act perfectly at the point of data entry?

**Reply by blelump** [reply]
To be clear, you would like to avoid this behavior as it will heavily impact how the whole Repo behaves. Let me provide a diagram that you can see below:
![Untitled Diagram](https://github.com/the-human-colossus-foundation/oca-spec/assets/425226/989f3e01-545d-46c3-90a4-81432d9edd64)

We have two histories leading to the same result (SAID4) because the list of steps used to create the Bundle ends up with the same canonical form of the Bundle. The difference is with SAID3 and SAID33, where the canonical form differs. What it means for the Repo custodian is that she would need to look for any change made in the Repo, which is impossible. Here she'd need to blacklist SAID4 and SAID33. Smuggling corrupted information into a Repo instance is easy. What we can do to prevent it is to add a reputation for those who either benefit or require it (with an authenticity layer). For those who don't care about reputation but demand uncorrupted information, again, the authenticity layer, that at least adds the possibility for governance to decide what to publish.

**Reply by carlyh-micb** [reply]
As I understand, SAID33 and SAID4 all have 'bad' data and all would be good to remove from search and FROM'ing. If SAID33 and all children were added to the blacklist, that means SAID4 would also be blacklisted, and even blacklisted on another history. Isn't this a good thing? Without doing a large search, we have blacklisted all the SAIDs that contain problematic information?

Reputation and authenticity layers are an excellent addition, but those are tools and technologies that are far from implementation. And they still don't make the information unavailable.

There will always be the chance of problematic data being added to a repo no matter how good the governance is surrounding it. People will make mistakes. Or information that wasn't a problem before, becomes a problem later.

In addition, there is GDPR and the right to be forgotten, which suggests we must be able to deal with this challenge.

Is there another solution to address. Is there an issue with DAG Vertex deletion?

**Reply by blelump** [reply]
> As I understand, SAID33 and SAID4 all have 'bad' data and all would be good to remove from search and FROM'ing. If SAID33 and all children were added to the blacklist, that means SAID4 would also be blacklisted, and even blacklisted on another history. Isn't this a good thing? Without doing a large search, we have blacklisted all the SAIDs that contain problematic information?
> 

We blacklist them because we knew them from the diagram above. Imagine the left DAG happens first. You blacklist SAID4. Sometime later, the malicious user pushed again with the right DAG. You now blacklist SAID33. In both cases, you must know they have happened, and right now, you have absolutely no idea about that. Also, due to how the canonical form is calculated, the malicious user can further rotate the order of commands in OCAFile and keep pushing infinitely. In other words, we can find endless ways to push information overlay with the text `Michal sucks`.

> Reputation and authenticity layers are an excellent addition, but those are tools and technologies that are far from implementation. And they still don't make the information unavailable.
> 
> There will always be the chance of problematic data being added to a repo no matter how good the governance is surrounding it. People will make mistakes. Or information that wasn't a problem before, becomes a problem later.
 
Looking for a shortcut, set up 2 Repos—one that is writable and the other that is read-only. Solely the custodian of the first Repo can push data to the 2nd Repo. 2nd Repo has solely reviewed bundles. Most of that can be established on the infrastructure level. One feat that has to be developed is a queue for reviewers to review pending bundles that wait in the writable Repo. Yet another option is to develop a feature that involves your university identity system and demands authorization before pushing to Repo.

In the first scenario, we involved a human that does the review upfront. We might further limit human interaction by employing AI that looks for malicious information.

> In addition, there is GDPR and the right to be forgotten, which suggests we must be able to deal with this challenge.

The right to be forgotten is about PII and sensitive information. In Repos (by design), we don't have any. In Repos, we solely keep schemas that humanly describe data.
